File: data_structures/queue/matlab/test_queue.m
% test_queue.m — キューのテストスクリプト

q = queue('new');
assert(queue(q, 'is_empty'));
assert(queue(q, 'size') == 0);

q = queue(q, 'enqueue', 10);
q = queue(q, 'enqueue', 20);
q = queue(q, 'enqueue', 30);
assert(queue(q, 'peek') == 10);
assert(queue(q, 'size') == 3);

[q, v] = queue(q, 'dequeue');
assert(v == 10);
[q, v] = queue(q, 'dequeue');
assert(v == 20);
assert(queue(q, 'peek') == 30);
[q, v] = queue(q, 'dequeue');
assert(v == 30);
assert(queue(q, 'is_empty'));

try
    empty_q = queue('new');
    queue(empty_q, 'dequeue');
    error('expected error not thrown');
catch e
    assert(strcmp(e.identifier, 'queue:underflow'));
end

for i = 1:50
    q = queue(q, 'enqueue', i);
end
assert(queue(q, 'size') == 50);
assert(queue(q, 'peek') == 1);

disp('All tests passed.');

File: data_structures/queue/cpp/test_queue.cpp
#include "queue.h"
#include <cstdlib>
#include <iostream>
#include <stdexcept>

namespace {
bool allPassed = true;
void pass(const std::string& n) { std::cout << "[PASS] " << n << "\n"; }
void fail(const std::string& n, const std::string& d) {
    std::cerr << "[FAIL] " << n << ": " << d << "\n"; allPassed = false;
}
void check(const std::string& n, bool c, const std::string& d = "assertion failed") {
    if (c) pass(n); else fail(n, d);
}
} // namespace

int main() {
    { Queue q; check("new is empty", q.is_empty()); }
    { Queue q; check("new size 0", q.size() == 0); }

    {
        Queue q;
        q.enqueue(1); q.enqueue(2); q.enqueue(3);
        check("FIFO dequeue1", q.dequeue() == 1);
        check("FIFO dequeue2", q.dequeue() == 2);
        check("FIFO dequeue3", q.dequeue() == 3);
        check("empty after all dequeue", q.is_empty());
    }

    {
        Queue q;
        q.enqueue(5);
        check("peek front", q.peek() == 5);
        check("peek no remove", q.size() == 1);
    }

    {
        Queue q;
        try { q.dequeue(); fail("dequeue empty throws", "no exception"); }
        catch (const std::underflow_error&) { pass("dequeue empty throws"); }
    }

    {
        Queue q;
        for (int i = 0; i < 100; ++i) q.enqueue(i);
        check("size 100", q.size() == 100);
        for (int i = 0; i < 100; ++i)
            if (q.dequeue() != i) { fail("large FIFO order", "mismatch at " + std::to_string(i)); break; }
        check("empty after 100", q.is_empty());
    }

    {
        Queue q;
        q.enqueue(1); q.dequeue();
        q.enqueue(2); q.dequeue();
        q.enqueue(3);
        check("wrap-around peek", q.peek() == 3);
        check("wrap-around size", q.size() == 1);
    }

    return allPassed ? EXIT_SUCCESS : EXIT_FAILURE;
}
